crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+            [first_item, .., last_item] => (first_item.start(), last_item.end()),
+            _ => panic!(
+                "We shouldn't be attempting an autofix if `__all__` has < 2 elements;
+                an `__all__` definition with 1 or 0 elements cannot be unsorted."
+            ),
+        };
+
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis,
+        //    up to the start of the first item in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    item in `__all__` up to and including the closing parenthesis.
+        //
+        // For example:
+        //
+        // ```python
+        // __all__ = [  # comment0
+        //   # comment1
+        //   "first item",
+        //   "last item"  # comment2
+        //   # comment3
+        // ]  # comment4
+        // ```
+        //
+        // - The prelude in the above example is the source code region
+        //   starting just before the opening `[` and ending just after `# comment0`.
+        //   `comment0` here counts as part of the prelude because it is on
+        //   the same line as the opening paren, and because we haven't encountered
+        //   any elements of `__all__` yet, but `comment1` counts as part of the first item,
+        //   as it's on its own line, and all comments on their own line are grouped
+        //   with the next element below them to make "items",
+        //   (an "item" being a region of source code that all moves as one unit
+        //   when `__all__` is sorted).
+        // - The postlude in the above example is the source code region starting
+        //   just after `# comment2` and ending just after the closing paren.
+        //   `# comment2` is part of the last item, as it's an inline comment on the
+        //   same line as an element, but `# comment3` becomes part of the postlude
+        //   because there are no items below it. `# comment4` is not part of the
+        //   postlude: it's outside of the source-code range considered by this rule,
+        //   and should therefore be untouched.
+        //
+        let newline = stylist.line_ending().as_str();
+        let start_offset = self.start();
+        let leading_indent = leading_indentation(locator.full_line(start_offset));
+        let item_indent = format!("{}{}", leading_indent, stylist.indentation().as_str());
+
+        let prelude =
+            multiline_dunder_all_prelude(first_item_start, newline, start_offset, locator);
+        let postlude = multiline_dunder_all_postlude(
+            last_item_end,
+            newline,
+            leading_indent,
+            &item_indent,
+            self.end(),
+            locator,
+        );
+
+        self.items
+            .sort_by_cached_key(|item| AllItemSortKey::from(item));
```

Comment:
Nit: We could consider to store `AllItemSortyKey` on the `DundarAllItem` if we're concerned about the cost. Although I don't think it's necessary because:

* This code path is only exercised for fixes, it's okay if they're somewhat slower
* We can change `AllItemSortKey` to us a `&str` to avoid cloning. The complexity then becomes `O(n)` but only for all capitalized strings.